### PR TITLE
Support python 3

### PR DESCRIPTION
**Pull request by zachmullen**
There is still quite a lot left to do on this branch, just wanted to publish it early so people could see the process.

### CODE REVIEW

**Comment by manthey** [comment]
We should eventually add tests to the travis matrix for python3 in addition
to the current python2.

On Thu, Apr 30, 2015 at 9:01 AM, Zach Mullen <notifications@github.com>
wrote:

> There is still quite a lot left to do on this branch, just wanted to
> publish it early so people could see the process.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/girder/girder/pull/816
> Commit Summary
>
>    - First round of core changes to support py3
>
> File Changes
>
>    - *M* .gitignore
>    <https://github.com/girder/girder/pull/816/files#diff-0> (1)
>    - *M* girder/api/api_main.py
>    <https://github.com/girder/girder/pull/816/files#diff-1> (2)
>    - *M* girder/api/rest.py
>    <https://github.com/girder/girder/pull/816/files#diff-2> (36)
>    - *M* girder/api/v1/user.py
>    <https://github.com/girder/girder/pull/816/files#diff-3> (2)
>    - *M* girder/events.py
>    <https://github.com/girder/girder/pull/816/files#diff-4> (11)
>    - *M* girder/models/model_base.py
>    <https://github.com/girder/girder/pull/816/files#diff-5> (15)
>    - *M* girder/models/password.py
>    <https://github.com/girder/girder/pull/816/files#diff-6> (12)
>    - *M* girder/models/token.py
>    <https://github.com/girder/girder/pull/816/files#diff-7> (7)
>    - *M* girder/utility/__init__.py
>    <https://github.com/girder/girder/pull/816/files#diff-8> (4)
>    - *M* girder/utility/config.py
>    <https://github.com/girder/girder/pull/816/files#diff-9> (6)
>    - *M* girder/utility/filesystem_assetstore_adapter.py
>    <https://github.com/girder/girder/pull/816/files#diff-10> (2)
>    - *M* girder/utility/gridfs_assetstore_adapter.py
>    <https://github.com/girder/girder/pull/816/files#diff-11> (2)
>    - *M* girder/utility/plugin_utilities.py
>    <https://github.com/girder/girder/pull/816/files#diff-12> (25)
>    - *M* girder/utility/system.py
>    <https://github.com/girder/girder/pull/816/files#diff-13> (4)
>    - *M* girder/utility/ziputil.py
>    <https://github.com/girder/girder/pull/816/files#diff-14> (7)
>    - *M* requirements-dev.txt
>    <https://github.com/girder/girder/pull/816/files#diff-15> (1)
>    - *M* requirements.txt
>    <https://github.com/girder/girder/pull/816/files#diff-16> (3)
>
> Patch Links:
>
>    - https://github.com/girder/girder/pull/816.patch
>    - https://github.com/girder/girder/pull/816.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/girder/girder/pull/816>.
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by zachmullen** [comment]
@manthey yep :)

**Comment by zachmullen** [comment]
If we wanted to keep our build matrix to just 2 entries, we could be unscientific and simply do

1. mongo 2.6 and python 2
2. mongo 3.x and python 3

It would keep our CI builds small, and I think it would be fairly obvious when a breakage occurs whether it is a breakage due to python2/3 issues versus mongo 2/3 issues.

**Comment by manthey** [comment]
This would miss a theoretical edge case of the mongo connector in pymongo
for different version pairings, but that should be handled by the pymongo
people.  Somehow this reminds me of using Taguchi's method for robust
design.

I'd want a comment in the build matrix saying that we are doing so (so that
someone looking at how we build doesn't think the mongo version is coupled
to the python version).

On Thu, Apr 30, 2015 at 9:19 AM, Zach Mullen <notifications@github.com>
wrote:

> If we wanted to keep our build matrix to just 2 entries, we could be
> unscientific and simply do
>
>    1. mongo 2.6 and python 2
>    2. mongo 3.x and python 3
>
> It would keep our CI builds small, and I think it would be fairly obvious
> when a breakage occurs whether it is a breakage due to python2/3 issues
> versus mongo 2/3 issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/girder/girder/pull/816#issuecomment-97777648>.
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by zachmullen** [comment]
:+1: We could also enable a 3 or 4 build process that happens only on master branch.

**Comment by zachmullen** [comment]
@mgrauer Can you go ahead and commit your fixes for girder_client onto this topic? It's pretty much just those failing and maybe one other intermittent test failure I need to sort out.

**Comment by zachmullen** [comment]
@mgrauer You'll also need to change https://github.com/girder/girder/blob/master/tests/cases/py_client/lib_test.py#L102 to mock `sys.moves.input` rather than our old `rawInput` wrapper.

**Comment by zachmullen** [comment]
Calling all @girder/developers , @girder/owners , and @cryos : this beast is ready for review. In addition to looking at the code, it would be great if people could run the system with python 2 and 3 and click through the web app to make sure everything still works.

**Comment by jbeezley** [comment]
I'm getting
```
$ pip -V
pip 6.1.1 from /Users/jbeezley/.pyenv/versions/girder-py34/lib/python3.4/site-packages (python 3.4)
$ pip install -r requirements.txt
...
Collecting boto==2.29.1 (from -r requirements.txt (line 5))
  Using cached boto-2.29.1.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/private/var/folders/1w/nmj4rhjd5nz7x0323syntl380000gn/T/pip-build-r83js1z0/boto/setup.py", line 37, in <module>
        from boto import __version__
      File "/private/var/folders/1w/nmj4rhjd5nz7x0323syntl380000gn/T/pip-build-r83js1z0/boto/boto/__init__.py", line 27, in <module>
        from boto.pyami.config import Config, BotoConfigLocations
    
    ----------------------------------------
    Command "python setup.py egg_info" failed with error code 1 in /private/var/folders/1w/nmj4rhjd5nz7x0323syntl380000gn/T/pip-build-r83js1z0/boto
```

Installing `requirements-dev.txt` installs `boto==2.38.0`, which works fine.  I think you just need to bump the boto version.

**Comment by jbeezley** [comment]
Sorry, I didn't checkout the branch.  My bad.

**Comment by manthey** [comment]
I've placed a running instance of Girder running under python3 as a demo
anyone can get to here:
  http://parakon:8009/girderpython3/


On Tue, May 5, 2015 at 7:04 PM, Zach Mullen <notifications@github.com>
wrote:

> Calling all @girder/developers
> <https://github.com/orgs/girder/teams/developers> , @girder/owners
> <https://github.com/orgs/girder/teams/owners> , and @cryos
> <https://github.com/cryos> : this beast is ready for review. In addition
> to looking at the code, it would be great if people could run the system
> with python 2 and 3 and click through the web app to make sure everything
> still works.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/girder/girder/pull/816#issuecomment-99255847>.
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by zachmullen** [comment]
@manthey :trophy: you rock

**Comment by manthey** [comment]
Something that may need to go in the docs:

When doing `pip install requirements-dev.txt`, I initially got an ASCII codec error.  It turns out that pip3 does weird things with locales when install Httpretty [1], which can be avoided by setting the locale to something that is UTF-8.  I did `locale-gen en_US.UTF-8` and set `LANG=en_US.utf8` in my environment.

[1] https://github.com/gabrielfalcao/HTTPretty/issues/108

File: docs/development.rst [hunk 1]
```diff
@@ -64,6 +64,15 @@ will override the defaults.
 Server Development
 ------------------
 
+All commits to the core python code must work in all versions of python 2.7 and above.
+Python code in plugins should also work in 2.7 and 3.x, but some plugins may depend
```
Comment (manthey) [inline docs/development.rst]: Do we want to claim 3.x?  I think we just want to say 3.4 and above, otherwise we are committing to testing with 3.0, etc.

Comment (zachmullen) [inline docs/development.rst]: I think we should claim 3.x. Since our code works in both 2.7 and 3.4 I think we'll be fairly safe.

**Comment by cryos** [comment]
This seems to work for my simple use cases, and the remaining issues are things I need to fix in my plugin code. This is looking great from my clicking around (I did not attempt to review the code very deeply and will leave that to the girder developers).

File: docs/development.rst (hunk 1 above)
Comment (jbeezley) [inline docs/development.rst]: You can probably safely assume it is compatible with 3.3, but 3.2 and below are not commonly supported among dependencies.

Comment (cpatrick) [inline docs/development.rst]: I'd favor being explicit about what versions we test with with the caveat that we aren't doing anything that should disallow 3.0-3.3.

Comment (zachmullen) [inline docs/development.rst]: That's reasonable, I will change it to say that.

**Comment by zachmullen** [comment]
@jbeezley @cpatrick I changed the docs.

@manthey I haven't run into the issue you mention about httpretty install (though I did run into several issues with it and trying to get it to play nicely with moto in both 2.7 and 3.4). If you think we should document the problem you ran into, could you please add that and push it to the head of this topic?

**Comment by manthey** [comment]
Using a quick test, it appears that we work just fine with Python 3.3.  We DON'T work with Python 3.2 at the moment, because the boto module that is installed with pip3.2 isn't actually Python 3 compatible.  There may be other issues with 3.2, that was just the first one that stopped things from working.

**Comment by manthey** [comment]
Excellent work @zachmullen :dizzy:.  This LGTM if you like the doc changes I just pushed.

**Comment by jbeezley** [comment]
:+1:  here.  `+582 −397` is not too bad all things considered.

**Comment by zachmullen** [comment]
Thanks all for the reviews

**Comment by cpatrick** [comment]
Late
[LGTM](http://lgtm.in/i/IG0Mh4Arn)

**Pull request merged**